File: python/starfield.py
#!/usr/bin/env python3
"""Small headless implementation of the shared CNA Starfield Courier contract."""
import sys


class Game:
    def __init__(self):
        self.reset()

    def reset(self):
        self.state, self.mask = 0, 0
        self.x = self.z = self.elapsed = self.score = self.hazard_x = 0.0
        self.hazard_direction = 1.0

    @staticmethod
    def distance2(ax, az, bx, bz):
        return (ax - bx) ** 2 + (az - bz) ** 2

    def update(self, seconds, move_x=0.0, move_z=0.0, boost=False, restart=False):
        if restart:
            self.reset()
            return
        if self.state in (2, 3):
            return
        if self.state == 0:
            self.state = 1
        dt = max(0.0, min(0.25, seconds))
        speed = 7.0 if boost else 4.0
        self.x = max(-10.0, min(10.0, self.x + max(-1.0, min(1.0, move_x)) * speed * dt))
        self.z = max(-10.0, min(10.0, self.z + max(-1.0, min(1.0, move_z)) * speed * dt))
        self.elapsed += dt
        self.hazard_x += self.hazard_direction * 2.0 * dt
        if self.hazard_x >= 7:
            self.hazard_x, self.hazard_direction = 7.0, -1.0
        elif self.hazard_x <= -7:
            self.hazard_x, self.hazard_direction = -7.0, 1.0
        for index, cell in enumerate(((-6, 0), (0, -5), (6, 0))):
            if not self.mask & (1 << index) and self.distance2(self.x, self.z, *cell) <= 0.81:
                self.mask |= 1 << index
                self.score += 100.0
        if self.distance2(self.x, self.z, self.hazard_x, 3) <= 3.0625 or self.elapsed >= 60:
            self.state = 3
        elif self.mask == 7 and self.distance2(self.x, self.z, 0, -9) <= 1.96:
            self.state, self.score = 2, self.score + 1000

    def snapshot(self):
        return self.state, self.mask, self.x, self.z, self.elapsed, self.score


def step(game, x, z, boost=True):
    game.update(0.25, x, z, boost)


def collect(game):
    for _ in range(4): step(game, -1, 0)
    for _ in range(4): step(game, 1, 0)
    for _ in range(3): step(game, 0, -1)
    for _ in range(4): step(game, 1, 0)


def scenario(name):
    game = Game()
    if name in ("collection", "win"):
        collect(game)
        if name == "win":
            for _ in range(3): step(game, -1, 0)
            for _ in range(2): step(game, 0, -1)
    elif name == "hazard":
        for _ in range(3): step(game, 0, 1, False)
    elif name == "loss":
        for _ in range(241): step(game, 0, 0, False)
    elif name == "restart":
        step(game, 1, 0)
        game.update(0, restart=True)
    elif name != "startup":
        raise SystemExit(f"unknown scenario: {name}")
    return game


if __name__ == "__main__":
    values = scenario(sys.argv[1] if len(sys.argv) > 1 else "startup").snapshot()
    print(values[0], values[1], *(f"{v:.4f}" for v in values[2:]))

File: ruby/starfield.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

class Game
  attr_reader :state, :mask, :x, :z, :elapsed, :score
  def initialize; reset; end
  def reset
    @state = 0; @mask = 0; @x = @z = @elapsed = @score = @hazard_x = 0.0; @hazard_direction = 1.0
  end
  def d2(ax, az, bx, bz); (ax - bx)**2 + (az - bz)**2; end
  def update(seconds, mx = 0, mz = 0, boost = false, restart = false)
    if restart then reset; return end
    return if [2, 3].include?(@state)
    @state = 1 if @state == 0; dt = [[seconds, 0].max, 0.25].min; speed = boost ? 7.0 : 4.0
    @x = [[@x + [[mx, -1].max, 1].min * speed * dt, -10].max, 10].min
    @z = [[@z + [[mz, -1].max, 1].min * speed * dt, -10].max, 10].min; @elapsed += dt
    @hazard_x += @hazard_direction * 2 * dt
    if @hazard_x >= 7 then @hazard_x = 7; @hazard_direction = -1 end
    if @hazard_x <= -7 then @hazard_x = -7; @hazard_direction = 1 end
    [[-6, 0], [0, -5], [6, 0]].each_with_index do |cell, i|
      bit = 1 << i; if (@mask & bit).zero? && d2(@x, @z, *cell) <= 0.81 then @mask |= bit; @score += 100 end
    end
    if d2(@x, @z, @hazard_x, 3) <= 3.0625 || @elapsed >= 60 then @state = 3
    elsif @mask == 7 && d2(@x, @z, 0, -9) <= 1.96 then @state = 2; @score += 1000 end
  end
end
def step(g, x, z, boost = true); g.update(0.25, x, z, boost); end
def collect(g); 4.times { step(g, -1, 0) }; 4.times { step(g, 1, 0) }; 3.times { step(g, 0, -1) }; 4.times { step(g, 1, 0) }; end
name = ARGV[0] || 'startup'; g = Game.new
case name
when 'collection', 'win'; collect(g); if name == 'win' then 3.times { step(g, -1, 0) }; 2.times { step(g, 0, -1) } end
when 'hazard'; 3.times { step(g, 0, 1, false) }
when 'loss'; 241.times { step(g, 0, 0, false) }
when 'restart'; step(g, 1, 0); g.update(0, 0, 0, false, true)
when 'startup'
else exit 2
end
puts format('%d %d %.4f %.4f %.4f %.4f', g.state, g.mask, g.x, g.z, g.elapsed, g.score)